language_service/src/lib.rs

```diff
@@ -121,27 +162,81 @@ impl<'a> LanguageService<'a> {
     /// for the document, typically when the document is first opened in the editor.
     /// It should also be called whenever the source code is updated.
     ///
+    /// This is the "entry point" for the language service's logic, after its constructor.
+    ///
     /// LSP: textDocument/didOpen, textDocument/didChange
-    pub fn update_document(&mut self, uri: &str, version: u32, text: &str) {
+    #[async_recursion(?Send)]
+    pub async fn update_document(&mut self, uri: &str, version: u32, text: &str) {
+        if self.currently_updating {
+            self.pending_update = Some(PendingUpdate {
```

Comment:
I think you need this to be a queue, not just a single item, otherwise you may miss edits. We could skip multiple updates to the same `uri` though.